sklearn/_loss/_loss.pyx

```diff
@@ -0,0 +1,1780 @@
+                n_samples, schedule='static', nogil=True, num_threads=n_threads
+            ):
+                loss[i] = closs_absolute_error(y_true[i], raw_prediction[i])
+        else:
+            for i in prange(
+                n_samples, schedule='static', nogil=True, num_threads=n_threads
+            ):
+                loss[i] = (sample_weight[i]
+                    * closs_absolute_error(y_true[i], raw_prediction[i]))
+
+        return np.asarray(loss)
+
+    def _gradient(
+        self,
+        Y_DTYPE_C[::1] y_true,
+        Y_DTYPE_C[::1] raw_prediction,
+        Y_DTYPE_C[::1] sample_weight,
+        G_DTYPE_C[::1] gradient,
+        int n_threads=1
+    ):
+        cdef:
+            int i
+            int n_samples = y_true.shape[0]
+
+        if sample_weight is None:
+            for i in prange(
+                n_samples, schedule='static', nogil=True, num_threads=n_threads
+            ):
+                gradient[i] = cgradient_absolute_error(y_true[i], raw_prediction[i])
+        else:
+            for i in prange(
+                n_samples, schedule='static', nogil=True, num_threads=n_threads
+            ):
+                gradient[i] = (
+                    sample_weight[i]
+                    * cgradient_absolute_error(y_true[i], raw_prediction[i])
+                )
+
+        return np.asarray(gradient)
+
+    def _gradient_hessian(
+        self,
+        Y_DTYPE_C[::1] y_true,
+        Y_DTYPE_C[::1] raw_prediction,
+        Y_DTYPE_C[::1] sample_weight,
+        G_DTYPE_C[::1] gradient,
+        G_DTYPE_C[::1] hessian,
+        int n_threads=1
+    ):
+        cdef:
+            int i
+            int n_samples = y_true.shape[0]
+            double2 dbl2
+
+        if sample_weight is None:
+            for i in prange(
+                n_samples, schedule='static', nogil=True, num_threads=n_threads
+            ):
+                dbl2 = cgrad_hess_absolute_error(y_true[i], raw_prediction[i])
+                gradient[i] = dbl2.val1
+                hessian[i] = dbl2.val2
+        else:
+            for i in prange(
+                n_samples, schedule='static', nogil=True, num_threads=n_threads
+            ):
+                dbl2 = cgrad_hess_absolute_error(y_true[i], raw_prediction[i])
+                gradient[i] = sample_weight[i] * dbl2.val1
+                hessian[i] = sample_weight[i] * dbl2.val2
+
+        return np.asarray(gradient), np.asarray(hessian)
+
+
+cdef class cPinballLoss(cLossFunction):
+    """Quantile Loss aka Pinball Loss with identity link.
+
+    Domain:
+    y_true and y_pred all real numbers
+    quantile in (0, 1)
+
+    Link:
+    y_pred = raw_prediction
+
+    Note: 2 * cPinballLoss(quantile=0.5) equals cAbsoluteError()
+    """
+
+    def __init__(self, quantile):
+        self.quantile = quantile
+
+    cdef double closs(self, double y_true, double raw_prediction) nogil:
+        return closs_pinball_loss(y_true, raw_prediction, self.quantile)
```

Comment:
It does not look like `closs` is being called anywhere. Where do you expect this function to be used?

Reply:
This is expected to be used in SGD, see https://github.com/scikit-learn/scikit-learn/blob/aa1fba991488676321baf46627259dfb669caf89/sklearn/linear_model/_sgd_fast.pyx#L52